src/Build/Globbing/MSBuildGlob.cs

```diff
@@ -0,0 +1,271 @@
+
+            _globRoot = globRoot;
+            _fileSpec = fileSpec;
+            _matchFileExpression = matchFileExpression;
+            _needsRecursion = needsRecursion;
+
+            // compile the regex since it's expected to be used multiple times
+            _regex =
+                new Lazy<Regex>(
+                    () => new Regex(_matchFileExpression, FileMatcher.DefaultRegexOptions | RegexOptions.Compiled),
+                    true);
+
+            //// todo: EngineFileUtilities.GetFileListUnescaped parses the regex again.
+            //_expandedFiles = new Lazy<string[]>(
+            //    () => EngineFileUtilities.GetFileListUnescaped(_globRoot, _fileSpec),
+            //    true);
+        }
+
+        /// <inheritdoc />
+        public bool IsMatch(string stringToMatch)
+        {
+            ErrorUtilities.VerifyThrowArgumentNull(stringToMatch, nameof(stringToMatch));
+
+            if (FileUtilities.PathIsInvalid(stringToMatch))
+            {
+                return false;
+            }
+
+            var normalizedString = NormalizeMatchInput(stringToMatch);
+
+            return _regex.Value.IsMatch(normalizedString);
+        }
+
+        ///// <inheritdoc />
+        //public IEnumerable<string> Expand()
+        //{
+        //    return _expandedFiles.Value;
+        //}
+
+        /// <summary>
+        ///     Similar to <see cref="IsMatch" /> but also provides the match groups for the glob parts
+        /// </summary>
+        /// <param name="stringToMatch"></param>
+        /// <returns></returns>
+        public MatchInfoResult MatchInfo(string stringToMatch)
+        {
+            ErrorUtilities.VerifyThrowArgumentNull(stringToMatch, nameof(stringToMatch));
+
+            if (FileUtilities.PathIsInvalid(stringToMatch) ||
+                !IsLegal)
+            {
+                return MatchInfoResult.Empty;
+            }
+
+            var normalizedInput = NormalizeMatchInput(stringToMatch);
+
+            bool isMatch;
+            string fixedDirectoryPart, wildcardDirectoryPart, filenamePart;
+            FileMatcher.GetRegexMatchInfo(
+                normalizedInput,
+                _regex.Value,
+                out isMatch,
+                out fixedDirectoryPart,
+                out wildcardDirectoryPart,
+                out filenamePart);
+
+            return new MatchInfoResult(isMatch, fixedDirectoryPart, wildcardDirectoryPart, filenamePart);
+        }
+
+        private string NormalizeMatchInput(string stringToMatch)
+        {
+            var rootedInput = Path.Combine(_globRoot, stringToMatch);
+            var normalizedInput = FileUtilities.GetFullPathNoThrow(rootedInput);
+
+            // Degenerate case when the string to match is empty.
+            // Ensure trailing slash because the fixed directory part has a trailing slash.
+            if (stringToMatch == string.Empty)
+            {
+                normalizedInput = normalizedInput + Path.DirectorySeparatorChar;
```

Comment:
EnsureTrailing slash does not work with empty strings, and GetFullPathNoThrow does not add a trailing slash. Moreover, if  stringToMatch has illegal path characters GetFullPathNoThrow does not do anything.